File: Problem174_range-sum-of-bst.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int rangeSumBST(TreeNode root, int L, int R) {
        int ans = 0;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);
        while(!stack.isEmpty()) {
            TreeNode node = stack.pop();
            if(node != null) {
                if(L <= node.val && node.val <= R)
                    ans += node.val;
                if(L < node.val)
                    stack.push(node.left);
                if(node.val < R)
                    stack.push(node.right);
            }
        }
        return ans;
    }
}

File: Problem-174 Range Sum of BST.py
# 938. Range Sum of BST
# https://leetcode.com/problems/range-sum-of-bst/

# Logic: We check the node's value with the range. As it is a BST, 
# the left side would be strictly less than and right side will be 
# strictly greater than the parent. If the parent is greater than 
# low only then we check the left subtree. Similarly if the parent 
# is less than high, only then we check the right subtree.

# Time Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def _helper(self, node, low, high):
        if not node:
            return 0
        
        s = 0
        if node.val >= low:
            s += self._helper(node.left, low, high)
        
        if node.val <= high:
            s += self._helper(node.right, low, high)
        
        if low <= node.val <= high:
            s += node.val
        
        return s
    
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        return self._helper(root, low, high)